exception-handling/practice-exception-handling.md

```diff
@@ -0,0 +1,424 @@
+
+An example of a working implementation:
+
+```python
+def is_phone_num_valid(phone_num):
+    if type(phone_num) is not str:
+        raise ValueError
+    return True
+```
+
+### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- Question 3 -->
+
+<!-- prettier-ignore-start -->
+### !challenge
+
+* type: code-snippet
+* language: python3.6
+* id: VLwvhu
+* title: Exception Handling
+
+### !question
+
+Given these tests, refactor this function so it will pass. Use `try ... except` syntax.
+
+- If no error is raised, return `cost_per_person`
+- If a `ZeroDivsionError` is raised, catch it, and return `0`
+
+```python
+def test_split_cost_evenly():
+    cost = split_cost_evenly(100, 4, .5)
+    assert cost == pytest.approx(37.5)
+
+
+def test_caught_zerodivisionerror_returns_zero():
+    cost = split_cost_evenly(100, 0, .5)
+    assert cost == 0
+```
+
+### !end-question
+### !placeholder
+
+```python
+def split_cost_evenly(total_cost, num_of_people, tip_percentage):
+    cost_with_tip = total_cost * (1 + tip_percentage)
+    cost_per_person = cost_with_tip / num_of_people
+    return cost_per_person
+```
+### !end-placeholder
+### !tests
+```python
+import unittest
+from main import *
+
+class TestChallenge(unittest.TestCase):
+    def test_split_cost_evenly(self):
+        cost = split_cost_evenly(100, 4, .5)
+        self.assertAlmostEqual(cost, 37.5)
+
+    def test_caught_zerodivisionerror_returns_zero(self):
+        cost = split_cost_evenly(100, 0, .5)
+        self.assertEqual(cost, 0)
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+def split_cost_evenly(total_cost, num_of_people, tip_percentage):
+    try:
+        cost_with_tip = total_cost * (1 + tip_percentage)
+        cost_per_person = cost_with_tip / num_of_people
+        return cost_per_person
+    except ZeroDivisionError:
+        return 0
+```
+
+### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- Question 4 -->
+
+<!-- prettier-ignore-start -->
+### !challenge
+
+* type: code-snippet
+* language: python3.6
+* id: j9kUX9
+* title: Exception Handling
+
+### !question
+
+Given these tests, refactor **_only_** the **`get_celebration_treat`** function so it will pass. Use `try ... except` syntax.
+
+In `get_celebration_treat`...
+
+- If no error is raised, then `get_celebration_treat` returns `"celebration {thirteenth_donut}!"`, where `thirteenth_donut` is the result of `grab_thirteenth_donut()`
+- If an `IndexError` is raised because of accessing something missing in `donuts`, then `get_celebration_treat` returns `"no donut :("`
+
+```python
+def test_get_celebration_treat():
+    bakers_dozen = [
+        "glazed donut", "old-fashioned", "jelly donut", "bear claw", "classic donut with pink frosting and sprinkles on top", "donut hole", "another glazed donut", "yet another glazed donut?", "cruller", "maple donut", "powdered sugar donut", "doughnut", "thirteenth donut"
```

Comment:
How is it that "Bismark" doesn't make the list of donuts?!?!?!?!?

Reply:
_That's_ the name of those?!?!?!?!?!? 😆